### ISSUE TITLE
In certain cases Observable could not be cancelled

### ISSUE THREAD

**Issue by YulawOne** [monix/monix#321]
```scala
  implicit val scheduler = Scheduler(ExecutionContext.global)

  val source = Observable(1).flatMap(_ => Observable.interval(1.seconds).map(_ => "tic"))

  val task = source.foreachL(println).runAsync

  scheduler.scheduleOnce(1, SECONDS, { () =>
    println("Cancelling")
    task.cancel()
  })

  Await.result(task, Duration.Inf)

```

Cancellation for Observable.interval will not be happen. I believe that due following code:

monix.reactive.internal.operators.ConcatMapObservable.ConcatMapSubscriber#signalFinish
```scala
    private def signalFinish(ex: Option[Throwable]): Unit =
      stateRef.getAndSet(WaitComplete(ex)) match {
        case WaitOnNextChild(_) =>
          // In this state we know we have no active task,
          // and we know that the last `onNext` was triggered
          // (but its future not necessarily finished)
          // so we can just signal the completion event
          if (ex.isEmpty) sendOnComplete() else out.onError(ex.get)

        case Active(_) | WaitComplete(_) | Cancelled =>
          // We either have an active task, or this is a duplicate
          // event, hence we cannot signal the completion signal here.
          ()

        case WaitOnActiveChild =>
          // Something is screwed up in our state machine :-(
          reportInvalidState(WaitOnActiveChild, "signalFinish")
      }
```

If I correctly understand, that this fragment is called after Observable(1) completion and current Active(cancellable) will be substituted with WaitComplete(ex), that will make impossible to cancel observable in future

**Comment by alexandru** [comment]
@YulawOne you are right 😒

This was due to a recent change to optimize `concatMap` and unfortunately this wasn't caught by any tests.

I'll do an emergency release tomorrow.

**Comment by alexandru** [comment]
@YulawOne you pinpointed the problem, that's impressive :-) 

Sorry for any inconvenience this may have caused.

**Comment by YulawOne** [comment]
@alexandru Thanks and thank you for fast response.

**Comment by alexandru** [comment]
@YulawOne same logic was applied in `mapTask` and that had to be fixed as well. Also added tests to guard against regressions.

Thanks a lot for the report. I'm postponing the release for tomorrow since I have another issue I want to solve.

**Comment by alexandru** [comment]
@YulawOne sorry for the wait, this was fixed in today's release 2.2.2
